elfpy/data/contract_interface.py

```diff
@@ -126,46 +186,84 @@ def get_event_object(
+        return None
+    function_outputs = function_abi.get("outputs")
+    if function_outputs is None:
+        logging.warning("function abi does not specify outputs")
+        return None
+    if not isinstance(function_outputs, Sequence):
+        logging.warning("function abi outputs are not a sequence")
+        return None
+    if len(function_outputs) > 1:  # multiple unnamed vars were returned
+        return_names_and_types = []
+        for output in function_outputs:
+            return_names_and_types.append(_get_name_and_type_from_abi(output))
+    if (
+        function_outputs[0].get("type") == "tuple" and function_outputs[0].get("components") is not None
+    ):  # multiple named outputs were returned in a struct
+        abi_components = function_outputs[0].get("components")
+        if abi_components is None:
+            logging.warning("function abi output componenets are not a included")
+            return None
+        return_names_and_types = []
+        for component in abi_components:
+            return_names_and_types.append(_get_name_and_type_from_abi(component))
+    else:  # final condition is a single output
+        return_names_and_types = [_get_name_and_type_from_abi(function_outputs[0])]
+    return return_names_and_types
+
+
+def smart_contract_read(contract: Contract, function_name: str, *fn_args, **fn_kwargs) -> dict[str, Any]:
+    """Return from a smart contract read call
+
+    .. todo::
+        function to recursively find component names & types
+        function to dynamically assign types to output variables
+            would be cool if this also put stuff into FixedPoint
     """
-    if request_kwargs is None:
-        request_kwargs = {}
-    provider = Web3.HTTPProvider(ethereum_node, request_kwargs)
-    web3 = Web3(provider)
-    web3.middleware_onion.inject(geth_poa.geth_poa_middleware, layer=0)
-    return web3
+    # get the callable contract function from function_name & call it
+    function: ContractFunction = contract.get_function_by_name(function_name)(*fn_args)  # , **fn_kwargs)
+    return_values = function.call(**fn_kwargs)
+    if not isinstance(return_values, Sequence):
+        return_values = [return_values]
+    if contract.abi:  # not all contracts have an associated ABI
+        return_names_and_types = contract_function_abi_outputs(contract.abi, function_name)
+        if return_names_and_types is not None:
+            if len(return_names_and_types) != len(return_values):
+                raise AssertionError(f"{len(return_names_and_types)=} must equal {len(return_values)=}.")
+            function_return_dict = dict(
+                (var_name_and_type[0], var_value)
+                for var_name_and_type, var_value in zip(return_names_and_types, return_values)
+            )
+            return function_return_dict
+    return {f"var_{idx}": value for idx, value in enumerate(return_values)}
+
+
+def smart_contract_transact(
```

Comment:
todo for tests again.  sorry to be broken record, want to get unit coverage on these low level things we're building.